ethereum/p2p/src/main/java/tech/pegasys/pantheon/ethereum/p2p/discovery/internal/RecursivePeerRefreshState.java

```diff
@@ -0,0 +1,167 @@
+/*
+ * Copyright 2018 ConsenSys AG.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
+ * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+package tech.pegasys.pantheon.ethereum.p2p.discovery.internal;
+
+import static tech.pegasys.pantheon.ethereum.p2p.discovery.internal.PeerDistanceCalculator.distance;
+
+import tech.pegasys.pantheon.ethereum.p2p.peers.Peer;
+import tech.pegasys.pantheon.ethereum.p2p.peers.PeerBlacklist;
+import tech.pegasys.pantheon.util.bytes.BytesValue;
+
+import java.time.Duration;
+import java.time.Instant;
+import java.util.ArrayList;
+import java.util.Comparator;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.SortedSet;
+import java.util.TreeSet;
+
+class RecursivePeerRefreshState {
+  private final int CONCURRENT_REQUEST_LIMIT = 3;
+  private final BytesValue target;
+  private final PeerBlacklist peerBlacklist;
+  private final BondingAgent bondingAgent;
+  private final NeighborFinder neighborFinder;
+  private final HashMap<Peer, Integer> anteMap;
+  private final SortedSet<Map.Entry<Peer, Integer>> distanceSortedPeers;
+  private final List<OutstandingRequest> outstandingRequestList;
```

Comment:
Given that we want to check if a peer ID is in this list it's probably better to use a `Map<BytesValue, OutstandingRequest` or possibly just `Map<BytesValue, Instant>` if the creation time is the only thing we're tracking.  I think I may have led you astray on this one, sorry. :)

Reply:
In the new new version of the timeout task, i.e. `executeTimeoutEvaluation()`, it likewise seems to me that having a bespoke object makes it more efficient

Reply:
You could use a `Map<BytesValue, OutstandingRequest>` to keep the custom class and still have things optimised for looking up the existence of a request, but given there are only ever 3 outstanding requests actually it's not worth it.